browser/ntp_sponsored_images/new_tab_page_branded_view_counter.cc

```diff
@@ -0,0 +1,206 @@
+  demo->logo_alt_text = "Technikke: For music lovers.";
+  demo->logo_company_name = "Technikke";
+  demo->logo_destination_url = "https://brave.com";
+  return demo;
+}
+
+class NewTabPageBrandedViewCounterFactory
+    : public BrowserContextKeyedServiceFactory {
+ public:
+  static NewTabPageBrandedViewCounter* GetForProfile(Profile* profile) {
+    return static_cast<NewTabPageBrandedViewCounter*>(
+        GetInstance()->GetServiceForBrowserContext(profile, true));
+  }
+
+  static NewTabPageBrandedViewCounterFactory* GetInstance() {
+    static base::NoDestructor<NewTabPageBrandedViewCounterFactory> instance;
+    return instance.get();
+  }
+  NewTabPageBrandedViewCounterFactory() : BrowserContextKeyedServiceFactory(
+        "NewTabPageBrandedViewCounter",
+        BrowserContextDependencyManager::GetInstance()) {}
+  ~NewTabPageBrandedViewCounterFactory() override = default;
+
+  NewTabPageBrandedViewCounterFactory(
+      const NewTabPageBrandedViewCounterFactory&) = delete;
+  NewTabPageBrandedViewCounterFactory& operator=(
+      const NewTabPageBrandedViewCounterFactory&) = delete;
+
+ private:
+  // BrowserContextKeyedServiceFactory:
+  KeyedService* BuildServiceInstanceFor(
+      content::BrowserContext* browser_context) const override {
+    NewTabPageBrandedViewCounter* instance = new NewTabPageBrandedViewCounter(
+        Profile::FromBrowserContext(browser_context));
+    return instance;
+  }
+  content::BrowserContext* GetBrowserContextToUse(
+      content::BrowserContext* context) const override {
+    return chrome::GetBrowserContextRedirectedInIncognito(context);
+  }
+};
+
+}  // namespace
+
+// static
+NewTabPageBrandedViewCounter* NewTabPageBrandedViewCounter::GetForProfile(
+    Profile* profile) {
+  if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaper))
+    return NewTabPageBrandedViewCounterFactory::GetForProfile(profile);
+  return nullptr;
+}
+
+NewTabPageBrandedViewCounter::NewTabPageBrandedViewCounter(Profile* profile)
+    : profile_(profile) {
+  // If we have a wallpaper, store it as private var.
+  // Set demo wallpaper if a flag is set.
+  if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo)) {
+    current_wallpaper_ = GetDemoWallpaper();
+  } else {
+    NTPSponsoredImagesComponentManager* manager =
+        g_brave_browser_process->ntp_sponsored_images_component_manager();
+    manager->AddObserver(this);
+    // Check if we have real data
+    const auto optional_data = manager->GetLatestSponsoredImagesData();
+    if (optional_data) {
+      current_wallpaper_.reset(new NTPSponsoredImagesData(*optional_data));
+      model_.set_total_image_count(
+          current_wallpaper_->wallpaper_image_urls.size());
+    }
+  }
+  // Allow notification dismissal pref to be reset when rewards status changes.
+  auto* rewards_service_ =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (rewards_service_)
+    rewards_service_->AddObserver(this);
```

Comment:
shouldn't we be checking the current state of rewards service (on/off)?

Reply:
Do you mean the user has turned it off, or some other 'off' state?

Reply:
right, if it's already off

Reply:
No I don't think we need to check if it's already off. We only need to check when it changes in order to clear out the notifcation-dismissed pref.